core/src/main/java/com/linecorp/armeria/client/ArrayBasedEventLoopState.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+
+import io.netty.channel.EventLoop;
+
+final class ArrayBasedEventLoopState extends AbstractEventLoopState {
+
+    private final AbstractEventLoopEntry[] entries;
+    private final int maxNumEventLoops;
+    private int allActiveRequests;
+
+    ArrayBasedEventLoopState(List<EventLoop> eventLoops, int maxNumEventLoops,
+                             DefaultEventLoopScheduler scheduler) {
+        super(eventLoops, scheduler);
+        this.maxNumEventLoops = maxNumEventLoops;
+        entries = new AbstractEventLoopEntry[maxNumEventLoops];
+        if (eventLoops.size() == maxNumEventLoops) {
+            init(0);
+        } else {
+            init(scheduler().acquisitionStartIndex(maxNumEventLoops));
+        }
+    }
+
+    private void init(final int acquisitionStartIndex) {
+        final int initialEventLoopOffset = ThreadLocalRandom.current().nextInt(maxNumEventLoops);
+        for (int i = 0; i < maxNumEventLoops; ++i) {
+            final int nextIndex = (acquisitionStartIndex + (initialEventLoopOffset + i) % maxNumEventLoops) %
+                                  eventLoops().size();
+            entries[i] = new Entry(this, eventLoops().get(nextIndex), i);
+        }
+    }
+
+    private AbstractEventLoopEntry targetEntry() {
+        int minActiveRequest = Integer.MAX_VALUE;
+        int targetIndex = 0;
+        for (int i = 0; i < maxNumEventLoops; ++i) {
+            final AbstractEventLoopEntry e = entries[i];
```

Comment:
We may encounter the worse case more frequently if an event loop is acquired from 0 index. What do you think of `allActiveRequests` as a seed value to randomly select the starting index?

Reply:
I think it would be better to use `ThreadLocalRandom.current()` than to designate seeds directly.

Reply:
Isn't it okay because the order of event loops  in the `entries` is different from the order of EventLoopsGroups?

Reply:
> the order of event loops in the entries is different from the order of EventLoopsGroups?

I didn't understand the meaning of `the order of event loops in the entries`. Should we sequentially acquire the entries from 0 index?

If we scan the array from the 0 index to find an empty or the least active connection, the lower part of the array tends to be filled more than the higher part of the array. 

> I think it would be better to use ThreadLocalRandom.current() than to designate seeds directly.

`ThreadLocalRandom.current()` is also a good choice. However, generating a random value is quite an expensive operation. Would you mind running benchmarks for the two seed values?

Reply:
> If we scan the array from the 0 index to find an empty or the least active connection, the lower part of the array tends to be filled more than the higher part of the array.

Yes but, the differences in active requests between the lower part and the higher part will be mostly just one which is acceptable.
Also, the request sent first is likely to be terminated first so I think we can get benefit from iterating from the same starting point.

Reply:
Not sure but it seems possible that the usage of CPU cores could be skewed if the lower part of event loops has a higher probability of selection.

Reply:
Had a chat with @ikhoon and it's probably okay with the current implementation. We will revisit later if we have any problems with this.